sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,412 @@
+            else
+            {
+                key = this.KeyResolver?.Resolve(encryptionData.WrappedContentKey.KeyId)
+                    ?? throw EncryptionErrors.KeyNotFound(encryptionData.WrappedContentKey.KeyId);
+            }
+
+            return key.UnwrapKey(
+                encryptionData.WrappedContentKey.Algorithm,
+                encryptionData.WrappedContentKey.EncryptedKey);
+        }
+
+        /// <summary>
+        /// Returns the key encryption key for blob. First tries to get key encryption key from KeyResolver, then
+        /// falls back to IKey stored on this EncryptionPolicy.
+        /// </summary>
+        /// <param name="encryptionData">The encryption data.</param>
+        /// <returns>Encryption key as a byte array.</returns>
+        private async Task<Memory<byte>> GetContentEncryptionKeyAsync(EncryptionData encryptionData)
+        {
+            IKeyEncryptionKey key;
+
+            // If we already have a local key and it is the correct one, use that.
+            if (encryptionData.WrappedContentKey.KeyId == this.LocalKey?.KeyId)
+            {
+                key = this.LocalKey;
+            }
+            // Otherwise, use the resolver.
+            else
+            {
+                key = await (this.KeyResolver?.ResolveAsync(encryptionData.WrappedContentKey.KeyId) ??
+                    throw EncryptionErrors.KeyNotFound(encryptionData.WrappedContentKey.KeyId)).ConfigureAwait(false);
+            }
+
+            return await key.UnwrapKeyAsync(
+                encryptionData.WrappedContentKey.Algorithm,
+                encryptionData.WrappedContentKey.EncryptedKey).ConfigureAwait(false);
+        }
+
+        #endregion
+
+        private static Stream WrapStream(Stream contentStream, byte[] contentEncryptionKey,
+            EncryptionData encryptionData, byte[] iv, bool noPadding)
+        {
+            switch (encryptionData.EncryptionAgent.EncryptionAlgorithm)
+            {
+                case ClientsideEncryptionAlgorithm.AES_CBC_256:
+                    using (AesCryptoServiceProvider aesProvider = new AesCryptoServiceProvider())
+                    {
+                        aesProvider.IV = iv ?? encryptionData.ContentEncryptionIV;
+                        aesProvider.Key = contentEncryptionKey;
+
+                        if (noPadding)
+                        {
+                            aesProvider.Padding = PaddingMode.None;
+                        }
+
+                        return new RollingBufferStream(new CryptoStream(contentStream, aesProvider.CreateDecryptor(), CryptoStreamMode.Read), 10 * Constants.MB);
+                    }
+                default:
+                    throw EncryptionErrors.BadEncryptionAlgorithm();
+            }
+        }
+
+        private static HttpRange ParseHttpRange(string serializedRange)
+        {
+            var rangeValues = serializedRange.Split('=')[1].Split(new char[] { '-' }, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
We should add a few more checks to avoid possible NullRefs.  And maybe move this into the HttpRange class.

Reply:
#8007

Reply:
I don't think that change will happen this release though because we're locked down on API changes in Azure.Core.  A few extra size checks here would still be good in the meantime.